internal/app/server/impl.go

```diff
@@ -0,0 +1,374 @@
+		for _, trailer := range responseDefinition.ResponseTrailers {
+			for _, val := range trailer.Value {
+				stream.ResponseTrailer().Add(trailer.Name, val)
+			}
+		}
+	}
+
+	// Convert the request to an Any so that it can be recorded in the payload
+	msgAsAny, err := asAny(req.Msg)
+	if err != nil {
+		return err
+	}
+	payload := initPayload(req.Header(), []*anypb.Any{msgAsAny})
+
+	for _, data := range responseDefinition.ResponseData {
+		payload.Data = data
+
+		resp := &v1alpha1.ServerStreamResponse{
+			Payload: payload,
+		}
+
+		if err := sendOnStream[v1alpha1.ServerStreamResponse](
+			ctx,
+			stream,
+			resp,
+			responseDefinition.ResponseDelayMs,
+		); err != nil {
+			return err
+		}
+
+	}
+	if responseDefinition.Error != nil {
+		return createError(responseDefinition.Error)
+	}
+	return nil
+}
+
+func (s *conformanceServer) BidiStream(
+	ctx context.Context,
+	stream *connect.BidiStream[v1alpha1.BidiStreamRequest, v1alpha1.BidiStreamResponse],
+) error {
+	var responseDefinition *v1alpha1.StreamResponseDefinition
+	fullDuplex := false
+	firstRecv := true
+	respNum := 0
+	var reqs []*anypb.Any
+	for {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		req, err := stream.Receive()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				// Reads are done, break the receive loop and send any remaining responses
+				break
+			}
+			return fmt.Errorf("receive request: %w", err)
+		}
+
+		// Record all requests received
+		msgAsAny, err := asAny(req)
+		if err != nil {
+			return err
+		}
+		reqs = append(reqs, msgAsAny)
+
+		// If this is the first message in the stream, save off the total responses we need to send
+		// plus whether this should be full or half duplex
+		if firstRecv {
+			responseDefinition = req.ResponseDefinition
+			fullDuplex = req.FullDuplex
+			firstRecv = false
+		}
+
+		// If fullDuplex, then send one of the desired responses each time we get a message on the stream
+		if fullDuplex {
+			if respNum >= len(responseDefinition.ResponseData) {
+				return connect.NewError(
+					connect.CodeAborted,
+					errors.New("received more requests than desired responses on a full duplex stream"),
+				)
+			}
+			payload := initPayload(stream.RequestHeader(), reqs)
+			payload.Data = responseDefinition.ResponseData[respNum]
+			resp := &v1alpha1.BidiStreamResponse{
+				Payload: payload,
+			}
+			err := sendOnStream[v1alpha1.BidiStreamResponse](
+				ctx,
+				stream,
+				resp,
+				responseDefinition.ResponseDelayMs,
+			)
+			if err != nil {
+				return err
+			}
+			respNum++
+			reqs = nil
+		}
+	}
+
+	// If we still have responses left to send, flush them now. This accommodates
+	// both scenarios of half duplex (we haven't sent any responses yet) or full duplex
+	// where the requested responses are greater than the total requests.
+	if respNum < len(responseDefinition.ResponseData) {
+		for i := respNum; i < len(responseDefinition.ResponseData); i++ {
```

Comment:
No need to introduce a new counter var.
```suggestion
		for ; respNum < len(responseDefinition.ResponseData); respNum++ {
```